### PR TITLE
Implement cancellation tokens

### PR DESCRIPTION
**Pull request by RedTaylor**

### CODE REVIEW

**Review by pepone — CHANGES_REQUESTED** [review]
There are a few demos in IceDiscovery/Manual/Glacier2/IceStorm that need similar fixes.

File: csharp/Ice/session/server/SessionFactory.cs [hunk 1]
```diff
@@ -18,7 +19,9 @@ public ISessionPrx Create(string name, Current current, CancellationToken cancel
             ISessionPrx collocProxy = proxy.Clone(endpoints: Array.Empty<Endpoint>());
 
             // Never close this connection from the client and turn on heartbeats with a timeout of 30s
-            current.Connection!.Acm = new Acm(TimeSpan.FromSeconds(30), AcmClose.Off, AcmHeartbeat.Always);
+            //      current.Connection!.IdleTimeout = TimeSpan.FromSeconds(30);
```
Comment (pepone) [inline csharp/Ice/session/server/SessionFactory.cs:22]: you can add a TODO to remember that this need to be fixed

File: csharp/Ice/invoke/server/Printer.cs [hunk 2]
```diff
@@ -10,7 +10,13 @@ namespace Demo
...
     public class Printer : IObject
     {
-        public ValueTask<OutgoingResponseFrame> DispatchAsync(IncomingRequestFrame request, Current current, CancellationToken cancel)
+        public ValueTask<OutgoingResponseFrame> DispatchAsync(
+
+            IncomingRequestFrame request,
+
+            Current current,
+
+            CancellationToken cancel)
```
Comment (pepone) [inline csharp/Ice/invoke/server/Printer.cs:19]: you should remove the empty lines here

File: csharp/Ice/multicast/HelloI.cs [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 
+using System;
+using System.Threading;
 using Demo;
 
 public class HelloI : HelloDisp_
 {
     public override void
-    sayHello(Ice.Current current)
+    sayHello(Ice.Current current, Ice.CancellationToken cancel)
```
Comment (bernardnormier) [inline csharp/Ice/multicast/HelloI.cs:10]: This Ice.CancellationToken does not look correct!

**Comment by pepone** [comment]
There are a few demos that are still missing the cancellation token updates

* Manual\printer
* Manual\simpleFilesystem
* IceDiscovery\hello
* Glacier2\callback
* IceStorm\clock
* IceGrid\icebox\helloservice

File: csharp/IceGrid/simple/server/Hello.cs [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System;
+using System.Threading.Tasks;
```
Comment (pepone) [inline csharp/IceGrid/simple/server/Hello.cs:4]:
```suggestion
using System.Threading;
```

File: csharp/IceDiscovery/replication/server/Hello.cs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
+using System.Threading.Tasks;
```
Comment (pepone) [inline csharp/IceDiscovery/replication/server/Hello.cs:3]:
```suggestion
using System.Threading;
```

File: csharp/IceBox/hello/helloservice/HelloService.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using ZeroC.Ice;
+using System.Threading.Tasks;
 using ZeroC.IceBox;
```
Comment (pepone) [inline csharp/IceBox/hello/helloservice/HelloService.cs:5]:
```suggestion
using System.Threading;
using ZeroC.Ice;
using ZeroC.IceBox;
```

File: csharp/Ice/interceptor/server/Authenticator.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System;
+using System.Threading;
 using System.Collections.Generic;
 using System.Security.Cryptography;
 using ZeroC.Ice;
```
Comment (pepone) [inline csharp/Ice/interceptor/server/Authenticator.cs:7]: The using directives should be sorted alphabetically
```suggestion
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Threading;
using ZeroC.Ice;
```

File: csharp/Ice/invoke/server/Printer.cs [hunk 8]
```diff
@@ -3,31 +3,35 @@
 using System;
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using System.Threading;
```
Comment (pepone) [inline csharp/Ice/invoke/server/Printer.cs:6]: Sort the using

```suggestion
using System.Threading;
using System.Threading.Tasks;
```

File: csharp/Ice/plugin/helloplugin/HelloPlugin.cs [hunk 9]
```diff
@@ -1,16 +1,17 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System.Threading.Tasks;
+using System.Threading;
```
Comment (pepone) [inline csharp/Ice/plugin/helloplugin/HelloPlugin.cs:4]:
```suggestion
using System.Threading;
using System.Threading.Tasks;
```

File: csharp/Ice/session/server/Session.cs [hunk 10]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System;
+using System.Threading;
 using System.Collections.Generic;
```
Comment (pepone) [inline csharp/Ice/session/server/Session.cs:5]:
```suggestion
using System.Collections.Generic;
using System.Threading;
```

File: csharp/Ice/session/server/SessionFactory.cs [hunk 11]
```diff
@@ -1,13 +1,15 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (pepone) [inline csharp/Ice/session/server/SessionFactory.cs:5]: `using System.Threading.Tasks;` is not required

```suggestion
using System.Threading;
```

File: csharp/Ice/throughput/server/Throughput.cs [hunk 12]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 
 using System;
+using System.Threading;
 using System.Collections.Generic;
```
Comment (pepone) [inline csharp/Ice/throughput/server/Throughput.cs:5]:
```suggestion
using System.Collections.Generic;
using System.Threading;
```

File: csharp/IceStorm/clock/subscriber/Subscriber.cs [hunk 13]
```diff
@@ -2,6 +2,7 @@
 
 using Demo;
 using System;
+using System.Threading;
 using System.Collections.Generic;
 using System.Configuration;
 using System.Diagnostics;
```
Comment (pepone) [inline csharp/IceStorm/clock/subscriber/Subscriber.cs:8]: Sort using directives 
```suggestion
using System.Collections.Generic;
using System.Configuration;
using System.Diagnostics;
using System.Threading;
```

**Review by pepone — COMMENTED** [review]
There are a couple of build errors and a few minor style issues, see comments.

**Review by pepone — APPROVED** [review]
Looks good!

**Review by bernardnormier — APPROVED** [review]
All good.

**Pull request merged**